src/github-handler/comment-handler/suggestion-patch-handler/fix-suggestion-hunks-handler.ts

```diff
@@ -0,0 +1,109 @@
+ * @param {Range[]} ranges a list of range objects with a start value and end value
+ * @param {number} value the value to search for in the range list
+ * @param {number} from the lower bound of the search indices. (optional)
+ */
+export function findRange(ranges: Range[], value: number, from = 0): number {
+  function binarySearch(lo: number, hi: number): number {
+    const mid = Math.floor((hi + lo) / 2);
+    if (lo > hi) {
+      return -1;
+    } else if (ranges[mid].start > value) {
+      return binarySearch(lo, mid - 1);
+    } else if (ranges[mid].end < value) {
+      return binarySearch(mid + 1, hi);
+    } else {
+      return mid;
+    }
+  }
+  return binarySearch(from, ranges.length - 1);
+}
+
+/**
+ * Get a subset of the hunks who exactly fit within a valid hunk scope.
+ * @param {Range[]} scope the list of valid ranges that a hunk can fit between
+ * @param {Hunk[]} suggestedHunks the hunks for the file to suggest
+ */
+function getValidSuggestionHunks(scope: Range[], suggestedHunks: Hunk[]) {
+  const validSuggestionHunks: Hunk[] = [];
+  let lastIndex = 0;
+  suggestedHunks.forEach(suggestedHunk => {
+    const validStartIndex = findRange(scope, suggestedHunk.oldStart, lastIndex);
+    const validEndIndex = findRange(scope, suggestedHunk.oldEnd, lastIndex);
+    lastIndex = validEndIndex;
+    // if the suggested range is not a subset of a single valid range, then skip
+    if (
+      validStartIndex < 0 ||
+      validEndIndex < 0 ||
+      validEndIndex !== validStartIndex
+    ) {
+      return;
+    }
+    validSuggestionHunks.push(suggestedHunk);
+  });
+  return validSuggestionHunks;
+}
+
+/**
+ * Remove all the hunks whose scope lies outside a valid scope/ranges.
+ * Remove the files that have no valid scope
+ * Remove the files that have no valid hunks
+ * @param {FileRanges} validRanges a map of each file's valid scope/ranges
+ * @param {FileHunks} suggestions the hunks for each file to suggest
+ */
+export function filterOutOfScopeHunks(
+  changeScope: FileRanges,
+  suggestions: FileHunks
+) {
+  suggestions.forEach((suggestedHunks: Hunk[], fileName: string) => {
+    const scope = changeScope.get(fileName);
+    if (!scope) {
+      suggestions.delete(fileName);
+    } else {
+      const validSuggestionHunks = getValidSuggestionHunks(
+        scope,
+        suggestedHunks
+      );
+      if (validSuggestionHunks.length) {
+        suggestions.set(fileName, validSuggestionHunks);
+      } else {
+        suggestions.delete(fileName);
+      }
+    }
+  });
+}
+
+/**
+ * Remove the suggestions whose filename is in the list of invalid files to suggest
+ * @param {string[]} invalidFiles a list of file names
+ * @param {FileHunks} suggestions the hunks for each file to suggest
+ */
+export function filterOutOfScopeFiles(
+  invalidFiles: string[],
+  suggestions: FileHunks
+) {
+  invalidFiles.forEach(file => {
+    if (suggestions.get(file) !== undefined) {
+      suggestions.delete(file);
```

Comment:
Generally, modifying this input data can be error prone for maintenance and should only be necessary if you really are memory constrained or have other good reasons to try to modify in place.

Reply:
Right. at this point this only deals with file ranges so the memory usage should be pretty small. Good point.